### PR TITLE
Add new UART hal (USART1 and USART3) and example

### PR DESCRIPTION
**Pull request by csagonero**
Signed-off-by: Cyril SAGONÉRO <csagonero@free.fr>

### CODE REVIEW

File: stm32f0_hal/examples/uart.rs [hunk 1]
```diff
@@ -1,23 +1,46 @@
 #![no_std]
 
+extern crate cortex_m;
 extern crate stm32f0_hal as hal;
+#[macro_use(interrupt)]
+extern crate stm32f0x2;
```
Comment (nicolas-rabault) [inline stm32f0_hal/examples/uart.rs:6]: For me you should not have to include this in hal examples.
I think hal is created to  hide stm32f0x2...
@pierre-rouanet any comment?

File: stm32f0_hal/examples/uart.rs [hunk 2]
```diff
@@ -1,23 +1,46 @@
...
+    rcc::init(); // Full Speed 48Mhz
+    loop {
+        if uart3.transmit_complete() {
+            uart3.send(0x55);
+        }
+    }
+}
 
-    loop {}
+interrupt!(USART1, receive_uart1);
```
Comment (nicolas-rabault) [inline stm32f0_hal/examples/uart.rs]: So related to my really first comment, we shouldn't see this interrupt thing.

File: stm32f0_hal/examples/uart.rs [hunk 3]
```diff
@@ -1,23 +1,46 @@
...
-    uart::setup(
-        6000000,
+fn main() {
+    let uart3 = uart::Uart::setup(
+        uart::Uarts::Uart3,
+        57600,
         uart::NBits::_8bits,
         uart::StopBits::_1b,
         uart::Parity::None,
-        |byte| uart::send(byte),
```
Comment (nicolas-rabault) [inline stm32f0_hal/examples/uart.rs:15]: We could pass an option function argument more here to specify interrupt function.

File: stm32f0_hal/src/uart.rs [hunk 4]
```diff
@@ -29,164 +30,295 @@ pub enum Parity {
     Odd,
 }
 
-pub fn setup<F>(baudrate: u32, nbits: NBits, nbstopbits: StopBits, parity: Parity, mut f: F)
```
Comment (nicolas-rabault) [inline stm32f0_hal/src/uart.rs]: I test the baudrate by setting it to 115200 but I get a 19323 baudrate output...

File: stm32f0_hal/examples/uart.rs (hunk 1 above)
Comment (pierre-rouanet) [reply stm32f0_hal/examples/uart.rs:6]: Agreed: it's now tracked in #25

File: stm32f0_hal/examples/uart.rs (hunk 3 above)
Comment (pierre-rouanet) [reply stm32f0_hal/examples/uart.rs:15]: Also tracked in #25

**Review by pierre-rouanet — COMMENTED** [review]
Look good enough for me at the moment. @nicolas-rabault ?

**Review by nicolas-rabault — APPROVED** [review]
You should probably skwash the last 3 commit but it's ok to me

**Pull request merged**